### PR TITLE
Data large Reference Implementation

### PR DESCRIPTION
**Pull request by suresh-srinivas**
Reference Implementation for Utilizing Large Pages for Data. This directory contains a simple reference implementation for automating the process of utilizing explicit huge pages for Data.

### CODE REVIEW

File: large_data/data-large-reference.cc [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        }
+    }
+    return -1; // nothing found
+
+}
+
+bool iodlr_hp_enabled() {
+    uint64_t val = iodlr_procmeminfo("HugePages_Total:");
+    if (val == 0 )
+        return false;
```
Comment (uttampawar) [inline large_data/data-large-reference.cc:105]: Simplify the code as.
if (val > 0) {
  return true; 
} 
return false;

File: large_data/data-large-reference.cc [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+
+void * iodlr_allocate(size_t s, size_t pgsz) {
+    int flags=0;
+    void *data;
+    if (pgsz ==  1048576* 1024)
+       flags = FLAGS_1G;
+    else if (pgsz == 2048*1024)
+       flags = FLAGS_2M;
+    else
```
Comment (uttampawar) [inline large_data/data-large-reference.cc:127]: set default value to flags=FLAGS_4K and remove "else" block.

File: large_data/data-large-reference.cc (hunk 1 above)
Comment (suresh-srinivas) [reply large_data/data-large-reference.cc]: Thanks updated.

File: large_data/data-large-reference.cc (hunk 2 above)
Comment (suresh-srinivas) [reply large_data/data-large-reference.cc]: Thanks updated

File: large_data/README.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  Returns true if Huge Pages is enabled on the system
+* size_t iodlr_get_hp_size()
+  Obtains the size of the explicit huge pages (2MB or 1GB) in bytes
+* void * iodlr_allocate(size_t s, size_t pgsz)
+  mmap size bytes with pgsz
+* void iodlr_deallocate(char *d, size_t s)
+  unmap
+* int MapStaticCodeToLargePages(hotstart, hotend)
+  Map region from hotstart to hotend to 2MB pages
+  Returns -1 if an error occurs while mapping
```
Comment (gabrielschulhof) [inline large_data/README.md:19]:
```suggestion
```

File: large_data/README.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  Returns the cycles (measured by rdtsc) to execute the test.
+
+* int64_t defaulttest()
+  Allocate a SIZE buffer with 4K Pages, zero fill it, and stride through it at 4K stride touching each byte of the 4K region.
+  Returns the cycles (measured by rdtsc) to execute the test.
+```
+
+# Infrastructure
+
+The Linux kernel in modern distros has large pages feature enabled by default. To check whether this is true for your kernel, use the following command and look for output lines containing “huge”:
```
Comment (gabrielschulhof) [inline large_data/README.md:35]:
```suggestion
The Linux kernel in modern distros has the large pages feature enabled by default. To check whether this is true for your kernel, use the following command and look for output lines containing “huge”:
```

File: large_data/README.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+HugePages_Free:        0
+HugePages_Rsvd:        0
+HugePages_Surp:        0
+Hugepagesize:       1048676 kB
+```
+The nonempty command output indicates that large page support is present, but the zero values indicate that no pages are configured for use. Also the kernel is configured to use 1GB pages
+
+Set the number of pages to be used.
+Each page is 1GB, so a value of 8 = 8GB.
+This command actually allocates memory, so this much amount of memory must be available.
```
Comment (gabrielschulhof) [inline large_data/README.md:48]:
```suggestion
This command actually allocates memory, so the requested amount of memory must be available.
```

File: large_data/data-large-reference.cc [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+    while(getline(ifs,map_line)){
+        istringstream iss(map_line);
+        string keyname;
+        uint64_t value;
+        string kb;
+        iss >> keyname;
+        iss >> value;
+        iss >> kb;
+//        cout << keyname << " " << value << "\n";
```
Comment (gabrielschulhof) [inline large_data/data-large-reference.cc:93]:
```suggestion
```

File: large_data/data-large-reference.cc [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+
+void iodlr_deallocate(char *d, size_t s) {
+    int i;
+    i = munmap (d, s);
+    assert (i != -1);
+}
+
+void zerofill(char *d, size_t s) {
+    memset (d, 0, s);
+}
```
Comment (gabrielschulhof) [inline large_data/data-large-reference.cc:142]: Do we really need this?

Comment (suresh-srinivas) [reply large_data/data-large-reference.cc:142]: the touch function will reference uninitialized data otherwise.

File: large_data/README.md (hunk 3 above)
Comment (suresh-srinivas) [reply large_data/README.md:19]: Thanks for reviewing the documentation as well!

**Pull request merged**